Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py

```diff
@@ -0,0 +1,159 @@
+import demistomock as demisto  # noqa: F401
+from CommonServerPython import *  # noqa: F401
+"""Script for identifying and recommending the most likely owners of a discovered service
+from those surfaced by Cortex ASM Enrichment.
+"""
+
+
+from typing import Dict, List, Any
+import traceback
+from itertools import groupby
+
+
+def score(owners: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    """
+    Owner score is the number of observations on that owner divided by the max number of observations
+    for any owner in the list
+
+    Expects `Count` key and replaces it with `Ranking Score`
+    """
+    if owners:
+        max_count = max(owner.get('Count', 1) for owner in owners)
+        for owner in owners:
+            count = owner.pop('Count', 1)
+            owner['Ranking Score'] = count / max_count
+    return owners
+
+
+def rank(owners: List[Dict[str, Any]], k: int = 5) -> List[Dict[str, Any]]:
+    """
+    Return up to k owners with the highest ranking scores
+    """
+    if k <= 0:
+        raise ValueError(f'Number of owners k={k} must be greater than zero')
+    return sorted(owners, key=lambda x: x['Ranking Score'])[:k]
+
+
+def justify(owners: List[Dict[str, str]]) -> List[Dict[str, str]]:
+    """
+    For now, `Justification` is the same as `Source`; in the future, will sophisticate
+    """
+    for owner in owners:
+        owner['Justification'] = owner.get('Source', '')
+    return owners
+
+
+def _canonicalize(owner: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Canonicalizes an owner dictionary and adds a deduplication key
+    `Canonicalization` whose value is either:
+        1. whitespace-stripped and lower-cased email, if email exists
+        2. whitespace-stripped and lower-cased name
+        3. empty string if neither exists
+    """
+    if owner.get('Email', ''):
+        owner['Canonicalization'] = owner['Email'].strip().lower()
+        owner['Email'] = owner['Canonicalization']
+    elif owner.get('Name', ''):
+        owner['Canonicalization'] = owner['Name'].strip().lower()
+        owner['Name'] = owner['Canonicalization']
+    else:
+        owner['Canonicalization'] = ''
+    return owner
+
+
+def canonicalize(owners: List[Dict[str, str]]) -> List[Dict[str, Any]]:
+    """
+    Calls _canonicalize on each owner.
+
+    Defensive handling of potentially missing or NoneType inputs
+    """
+    canonicalized = []
```

Comment:
Rather than forcing new values in the RHS, what if we drop the broken ones entirely and log that we're doing so?  No overwriting of values, and we canonicalize and keep anything that obeys the contract.

```
canonicalized = []
try:
  for owner in owners:
    try:
       canonicalized.append(_canonicalize(owner))
    except Exception as e:
       demisto.error(f"Unable to canonicalize {owner}: {e}")
except Exception as e:
   demisto.error(f"`owners` must be iterable: {e}")
return canonicalized
```